pep-0xxx.rst

```diff
@@ -0,0 +1,1149 @@
+                lowest_supported_version = self.lowest_supported_version
+
+            if highest_supported_version is _DEFAULT_VALUE:
+                highest_supported_version = self.highest_supported_version
+
+            if trust_store is _DEFAULT_VALUE:
+                trust_store = self.trust_store
+
+            if sni_callback is _DEFAULT_VALUE:
+                sni_callback = self.sni_callback
+
+            return self.__class__(
+                validate_certificates, certificate_chain, ciphers,
+                inner_protocols, lowest_supported_version,
+                highest_supported_version, trust_store, sni_callback
+            )
+
+
+
+Context
+~~~~~~~
+
+The ``Context`` abstract base class defines an object that allows configuration
+of TLS. It can be thought of as a factory for ``TLSWrappedSocket`` and
+``TLSWrappedBuffer`` objects.
+
+As much as possible implementers should aim to make these classes immutable:
+that is, they should prefer not to allow users to mutate their internal state
+directly, instead preferring to create new contexts from new TLSConfiguration
+objects. Obviously, the ABCs cannot enforce this constraint, and so they do not
+attempt to.
+
+The ``Context`` abstract base class has the following class definition::
+
+    TLSBufferObject = Union[TLSWrappedSocket, TLSWrappedBuffer]
+
+
+    class _BaseContext(metaclass=ABCMeta):
+        @abstractmethod
+        def __init__(self, configuration: TLSConfiguration):
+            """
+            Create a new context object from a given TLS configuration.
+            """
+
+        @property
+        @abstractmethod
+        def configuration(self) -> TLSConfiguration:
+            """
+            Returns the TLS configuration that was used to create the context.
+            """
+
+
+    class ClientContext(_BaseContext):
+        @abstractmethod
+        def wrap_socket(self, socket: socket.socket,
+                        auto_handshake=True: bool,
+                        server_hostname=None: Optional[str]) -> TLSWrappedSocket:
+            """
+            Wrap an existing Python socket object ``socket`` and return a
+            ``TLSWrappedSocket`` object. ``socket`` must be a ``SOCK_STREAM``
+            socket: all other socket types are unsupported.
+
+            The returned SSL socket is tied to the context, its settings and
+            certificates.
+
+            The parameter ``auto_handshake`` specifies whether to do the SSL
+            handshake automatically after doing a ``socket.connect()``, or
+            whether the application program will call it explicitly, by
+            invoking the ``TLSWrappedSocket.do_handshake()`` method. Calling
+            ``TLSWrappedSocket.do_handshake()`` explicitly gives the program
+            control over the blocking behavior of the socket I/O involved in
+            the handshake.
+
+            The optional parameter ``server_hostname`` specifies the hostname
```

Comment:
This is just documentation, but perhaps we should state that it is recommend that this be passed in if at all possible? AFAICT it's impossible for this to validate the hostname without it so it is good for more than just SNI. I would even go so far as to say maybe the default should not be ``None``, but should be a sentinel value that behaves as if ``None`` was passed, but which prints a warning asking for a ``server_hostname`` or an explicit ``None`` to silence the warning.